File: examples/android/VistreaDemoApp/app/src/main/kotlin/dev/vistrea/demo/ui/StoreCatalog.kt
package dev.vistrea.demo.ui

import kotlin.math.roundToInt

/**
 * The deterministic storefront catalog shared by the store navigation and
 * store search scenarios. Every value is a pure function of the item index,
 * so repeated launches, scrolls, and queries always observe identical
 * content. Pure Kotlin: plain JVM unit tests cover it directly.
 */
internal object StoreCatalog {
    /** One deterministic catalog entry; text is content, never identity. */
    data class Item(
        val index: Int,
        val title: String,
        val price: String,
    )

    private val ADJECTIVES = listOf(
        "Aurora", "Basalt", "Cedar", "Drift", "Ember",
        "Fjord", "Glacier", "Harbor", "Iris", "Juniper",
    )

    private val PRODUCTS = listOf("Lamp", "Chair", "Kettle", "Rug", "Vase")

    /**
     * Exactly 50 items: every adjective and product combination, in order.
     * The iOS Demo App generates the identical titles and prices from the
     * same formulas, so cross-platform runs observe the same content.
     */
    val items: List<Item> = (0 until ADJECTIVES.size * PRODUCTS.size).map { index ->
        Item(
            index = index,
            title = "${ADJECTIVES[index % ADJECTIVES.size]} ${PRODUCTS[index / ADJECTIVES.size]}",
            price = "$${19 + (index * 7) % 80}.00",
        )
    }

    /**
     * Case-insensitive substring filter over item titles. A blank query
     * restores the complete catalog, so clearing the search field always
     * reproduces the exact launch structure.
     */
    fun search(query: String): List<Item> {
        val needle = query.trim()
        if (needle.isEmpty()) {
            return items
        }
        return items.filter { it.title.contains(needle, ignoreCase = true) }
    }
}

/**
 * Pure whole-row snap math for the storefront catalog window. The visible
 * window only ever moves in whole-row steps, so the rendered tree is snapped
 * at every instant of a gesture, not just after settling.
 */
internal object SnapCatalogWindow {
    /** How many whole rows a drag of [dragDistancePx] shifts the window. */
    fun rowShift(dragDistancePx: Float, rowHeightPx: Int): Int {
        require(rowHeightPx > 0) { "Row height must be positive." }
        return (dragDistancePx / rowHeightPx).roundToInt()
    }

    /** The clamped first visible index after applying a drag to a window. */
    fun settledFirstIndex(
        startFirstIndex: Int,
        dragDistancePx: Float,
        rowHeightPx: Int,
        itemCount: Int,
        visibleRows: Int,
    ): Int {
        val maximumFirstIndex = maxOf(0, itemCount - visibleRows)
        return (startFirstIndex + rowShift(dragDistancePx, rowHeightPx))
            .coerceIn(0, maximumFirstIndex)
    }
}

File: examples/ios/VistreaDemoApp/Sources/StoreCatalog.swift
import Foundation

/// One deterministic storefront catalog entry. Item text is content, not
/// identity: rows derived from these items never carry stable node IDs.
struct StoreCatalogItem: Equatable, Sendable {
    let name: String
    let price: String
}

/// The deterministic catalog shared by `demo.store.navigation` (the snapping
/// shop list) and `demo.store.search` (the filterable result list). The data
/// is a pure function of fixed word tables, so every launch, capture, and
/// platform observes the same fifty items in the same order.
enum StoreCatalog {
    static let itemCount = 50

    private static let families = [
        "Aurora", "Basalt", "Cedar", "Drift", "Ember",
        "Fjord", "Glacier", "Harbor", "Iris", "Juniper",
    ]
    private static let kinds = ["Lamp", "Chair", "Kettle", "Rug", "Vase"]

    /// The featured item pinned above the scrolling shop catalog.
    static var featuredItem: StoreCatalogItem {
        items[0]
    }

    static let items: [StoreCatalogItem] = (0..<itemCount).map { index in
        StoreCatalogItem(
            name: "\(families[index % families.count]) \(kinds[(index / families.count) % kinds.count])",
            price: "$\(19 + (index * 7) % 80).00"
        )
    }

    /// Case-insensitive substring filter over item names. A blank query is
    /// treated as no query and returns the complete catalog.
    static func matches(query: String) -> [StoreCatalogItem] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return items
        }
        return items.filter { $0.name.localizedCaseInsensitiveContains(trimmed) }
    }
}

/// Deterministic text values behind the shared fixtures' `input_alias`
/// entries. Automation resolves an alias to the exact string it types, so the
/// mapping must stay stable across platforms and launches.
enum ScenarioInputAliases {
    static func text(for alias: String) -> String? {
        switch alias {
        case "VALID_NAME":
            "Ada Lovelace"
        case "QUERY_MATCHING":
            "Aurora"
        case "QUERY_UNMATCHED":
            "Obsidian"
        default:
            nil
        }
    }
}
